From 66018b18a32b6c060c3f882ebfbcf787e5c707f6 Mon Sep 17 00:00:00 2001
From: DJ Delorie <dj@redhat.com>
Date: Mon, 25 Sep 2017 19:43:09 -0400
Subject: nscd: Release read lock after resetting timeout.

	[BZ #22161]
	* nscd/netgroupcache.c (addinnetgrX): Release read lock after
	resetting timeout.

Reviewed-by: Jonathan Nieder <jrnieder@gmail.com>
Reviewed-by: Andreas Schwab <schwab@suse.de>

A lock is held by mempool_allocate() when CACHEABLE is true; we
must release this lock if we exit early.
---
 nscd/netgroupcache.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/nscd/netgroupcache.c b/nscd/netgroupcache.c
index c36159ccfa..23b46d9934 100644
--- a/nscd/netgroupcache.c
+++ b/nscd/netgroupcache.c
@@ -558,6 +558,8 @@ addinnetgrX (struct database_dyn *db, int fd, request_header *req,
       dh->timeout = timeout;
       dh->ttl = dataset->head.ttl;
       ++dh->nreloads;
+      if (cacheable)
+        pthread_rwlock_unlock (&db->lock);
       return timeout;
     }
 
-- 
2.30.2

